micro-rdk/src/common/conn/server.rs

```diff
@@ -296,40 +328,91 @@ where
+                        let _ = async_io::Timer::after(duration).await;
+                        let rguard = app_client.read().await;
+                        for app_client in rguard.as_ref().iter() {
+                            match task.invoke(app_client).await {
+                                Ok(None) => continue,
+                                Ok(Some(next_duration)) => {
+                                    duration = next_duration;
+                                }
+                                Err(e) => {
+                                    log::warn!(
+                                        "Periodic task {} failed with error {}",
+                                        task.name(),
+                                        e
+                                    );
+                                }
+                            }
+                        }
+                    }
+                })
+            })
+            .collect();
+
         loop {
             let _ = async_io::Timer::after(std::time::Duration::from_millis(300)).await;
 
-            if self.app_client.is_none() {
-                let conn = self.app_connector.connect().await.unwrap();
-                let cloned_exec = self.exec.clone();
-                let grpc_client = Box::new(
-                    GrpcClient::new(conn, cloned_exec, "https://app.viam.com:443")
+            {
+                let urguard = self.app_client.upgradable_read().await;
+                if urguard.is_none() {
+                    let conn = self.app_connector.connect().await.unwrap();
+                    let cloned_exec = self.exec.clone();
+                    let grpc_client = Box::new(
+                        GrpcClient::new(conn, cloned_exec, "https://app.viam.com:443")
+                            .await
+                            .unwrap(),
+                    );
+                    let app_client = AppClientBuilder::new(grpc_client, self.app_config.clone())
+                        .build()
                         .await
-                        .unwrap(),
-                );
-                let app_client = AppClientBuilder::new(grpc_client, self.app_config.clone())
-                    .build()
-                    .await
-                    .unwrap();
-                let _ = self.app_client.insert(app_client);
+                        .unwrap();
+                    let _ = AsyncRwLockUpgradableReadGuard::upgrade(urguard)
+                        .await
+                        .insert(app_client);
+                }
             }
 
             let sig = if let Some(webrtc_config) = self.webrtc_config.as_ref() {
                 let ip = self.app_config.get_ip();
-                let signaling = self.app_client.as_mut().unwrap().connect_signaling();
+                let rguard = self.app_client.read().await;
+                let signaling = rguard.as_ref().unwrap().initiate_signaling();
```

Comment:
let me know if this is too much for this commit, but I wonder if we can make the actual signaling server logic just another instance of `PeriodicAppClientTask`? It feels like it would be cleaner with the one glaring problem maybe being the complexity of also accomplishing the `app_client` invalidation in L. 475. Maybe `invoke` can also return whether the `app_client` needs to be invalidated?

Reply:
I think it is an interesting idea and it goes in the direction I think this is headed of decomposing `serve` into services, but I don't think I want to undertake it before getting you unblocked. Would a ticket be OK?

Reply:
yeah, I think a ticket would be perfectly fine